### PR TITLE
Add an argument "negative_prompt"

### PR DESCRIPTION
**Pull request by shirayu**
Add `negative_prompt` argument to customize unconditional input.

# Example

## Bouquet of Roses
Bouquet of Roses

## Bouquet of Roses (`negative_prompt`=`red rose`)
Bouquet of Roses (`negative_prompt`=`red rose`

### CODE REVIEW

**Comment by OrionFive** [comment]
Super useful against extra arms and wonky hands. Thanks for sharing this!

**Comment by leszekhanusz** [comment]
That's great!

Some remarks:
 * it seems you have some issues around `batch_size`, sometimes multiplying by `batch_size` twice and the code is not exactly the same in all the pipelines
 * Please use `TypeError` instead of `ValueError` when the type of something is not correct

**Comment by shirayu** [comment]
@leszekhanusz Thank you for the comments! I fixed them.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 1]
```diff
@@ -203,9 +206,21 @@ def __call__(
         do_classifier_free_guidance = guidance_scale > 1.0
         # get unconditional embeddings for classifier free guidance
         if do_classifier_free_guidance:
+            ucond_tokens: List[str]
+            if negative_prompt is None:
+                ucond_tokens = [""] * batch_size
+            elif type(prompt) is not type(negative_prompt):
+                raise TypeError("`negative_prompt` should be the same type to `prompt`.")
+            elif isinstance(negative_prompt, str):
+                ucond_tokens = [negative_prompt]
```
Comment (leszekhanusz) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:215]: Shouldn't you multiply by batch size here now ?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 2]
```diff
@@ -231,9 +234,21 @@ def __call__(
         do_classifier_free_guidance = guidance_scale > 1.0
         # get unconditional embeddings for classifier free guidance
         if do_classifier_free_guidance:
+            ucond_tokens: List[str]
+            if negative_prompt is None:
+                ucond_tokens = [""] * batch_size
+            elif type(prompt) is not type(negative_prompt):
+                raise TypeError("`negative_prompt` should be the same type to `prompt`.")
+            elif isinstance(negative_prompt, str):
+                ucond_tokens = [negative_prompt]
```
Comment (leszekhanusz) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:243]: Shouldn't you multiply by batch size here now ?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py [hunk 3]
```diff
@@ -264,9 +267,21 @@ def __call__(
         do_classifier_free_guidance = guidance_scale > 1.0
         # get unconditional embeddings for classifier free guidance
         if do_classifier_free_guidance:
+            ucond_tokens: List[str]
+            if negative_prompt is None:
+                ucond_tokens = [""] * batch_size
+            elif type(prompt) is not type(negative_prompt):
+                raise TypeError("`negative_prompt` should be the same type to `prompt`.")
+            elif isinstance(negative_prompt, str):
+                ucond_tokens = [negative_prompt]
```
Comment (leszekhanusz) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:276]: Shouldn't you multiply by batch size here now ?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_onnx.py [hunk 4]
```diff
@@ -81,9 +82,21 @@ def __call__(
         do_classifier_free_guidance = guidance_scale > 1.0
         # get unconditional embeddings for classifier free guidance
         if do_classifier_free_guidance:
+            ucond_tokens: List[str]
+            if negative_prompt is None:
+                ucond_tokens = [""] * batch_size
+            elif type(prompt) is not type(negative_prompt):
+                raise TypeError("`negative_prompt` should be the same type to `prompt`.")
+            elif isinstance(negative_prompt, str):
+                ucond_tokens = [negative_prompt]
```
Comment (leszekhanusz) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_onnx.py:91]: Shouldn't you multiply by batch size here now ?

**Comment by shirayu** [comment]
@leszekhanusz Thank you! I forgot to make those changes. I fixed them.

**Comment by WASasquatch** [comment]
I tried to add your modified pipelines to try them out, and I got the following error:

```
Traceback (most recent call last):

  File "/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py", line 3326, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)

  File ["<ipython-input-2-5204ba97f9a8>"](https://localhost:8080/#), line 1217, in <module>
    from diffusers import PNDMScheduler, LMSDiscreteScheduler, DDIMScheduler, DDPMScheduler

  File "/usr/local/lib/python3.7/dist-packages/diffusers/__init__.py", line 32, in <module>
    from .pipelines import DDIMPipeline, DDPMPipeline, KarrasVePipeline, LDMPipeline, PNDMPipeline, ScoreSdeVePipeline

  File "/usr/local/lib/python3.7/dist-packages/diffusers/pipelines/__init__.py", line 12, in <module>
    from .stable_diffusion import (

  File "/usr/local/lib/python3.7/dist-packages/diffusers/pipelines/stable_diffusion/__init__.py", line 32, in <module>
    from .pipeline_stable_diffusion_img2img import StableDiffusionImg2ImgPipeline

  File "/usr/local/lib/python3.7/dist-packages/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py", line 123
    self,
    ^
SyntaxError: non-default argument follows default argument
```

https://github.com/huggingface/diffusers/blob/0301e53b223fca349c45031eaf8144980e9b41fd/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py#L123

Do you have any idea?

**Comment by exo-pla-net** [comment]
Super excited for this!

**Comment by shirayu** [comment]
@anton-l @patrickvonplaten Could you review this?

**Comment by WASasquatch** [comment]
I've been playing with this on Easy Diffusion and it's been working great, and generations so much better. This is definitely a needed feature.

**Comment by Hafiidz** [comment]
@shirayu, is there any guide to get this running in my local version while waiting for the update to main library ya?

**Comment by OrionFive** [comment]
> @shirayu, is there any guide to get this running in my local version while waiting for the update to main library ya?

The easiest way is where your code downloads the git repo from huggingface to replace that with this branch from shirayu's fork.

**Comment by patrickvonplaten** [comment]
Is this related to Dream Booth cc @patil-suraj ?

**Comment by WASasquatch** [comment]
Schedulers no longer has a method `set_format()`

**Comment by pcuenca** [comment]
> Schedulers no longer has a method `set_format()`

Hi @WASasquatch! You are right, #651 should fix it. Sorry!

**Comment by keturn** [comment]
> Is this related to Dream Booth cc @patil-suraj ?

This is generally applicable to all the stable diffusion classifier-free guidance pipelines. I don't think there's any particular connection to DreamBooth.

**Comment by WASasquatch** [comment]
I hope this gets added soon, so it doesn't have to be maintained so long while they do other stuff.

**Comment by patrickvonplaten** [comment]
@patil-suraj could you please also take a look here?

**Comment by WASasquatch** [comment]
Getting the following error on negative prompts inpaint pipeline

```python
[<ipython-input-22-b6b8ede7b450>](https://localhost:8080/#) in diffuse_run()
    507                 if mask is not None:
    508                     with autocast("cuda"):
--> 509                         pipeout = pipe(prompt=PROMPT, negative_prompt=NEG_PROMPT, num_inference_steps=STEPS, init_image=init, mask_image=mask, strength=INIT_SCALE, guidance_scale=SCALE, generator=gen_seed)
    510                         image = pipeout["sample"][0]
    511                 else:

[/usr/local/lib/python3.7/dist-packages/torch/autograd/grad_mode.py](https://localhost:8080/#) in decorate_context(*args, **kwargs)
     25         def decorate_context(*args, **kwargs):
     26             with self.clone():
---> 27                 return func(*args, **kwargs)
     28         return cast(F, decorate_context)
     29 

[/usr/local/lib/python3.7/dist-packages/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py](https://localhost:8080/#) in __call__(self, prompt, init_image, mask_image, strength, num_inference_steps, guidance_scale, eta, negative_prompt, generator, output_type, return_dict)
    318 
    319             # masking
--> 320             init_latents_proper = self.scheduler.add_noise(init_latents_orig, noise, t)
    321             latents = (init_latents_proper * mask) + (latents * (1 - mask))
    322 

[/usr/local/lib/python3.7/dist-packages/diffusers/schedulers/scheduling_pndm.py](https://localhost:8080/#) in add_noise(self, original_samples, noise, timesteps)
    393             self.alphas_cumprod = self.alphas_cumprod.to(original_samples.device)
    394 
--> 395         if timesteps.device != original_samples.device:
    396             timesteps = timesteps.to(original_samples.device)
    397 

AttributeError: 'numpy.int64' object has no attribute 'device'
```

**Comment by keturn** [comment]
There have been a few things merged this week dealing with making sure the scheduler data is using pytorch tensors and the data type of timesteps.

Without knowing which of those two `device` references on line 395 is the problem, my _hunch_ would be is that the inpainting pipeline's `init_latents_orig` are numpy arrays, and they need to be torch tensors now?

The other option is that some wires got crossed between PRs on whether timestamps are numpy types or not.

The kinda weird thing, though, is it doesn't say `'numpy.ndarray' object has no attribute`, it says `'numpy.int64' object`, so it's just a single scalar and not an array? That makes me second-guess my earlier hunches and think it's something else again.

**Comment by WASasquatch** [comment]
Users started getting new errors later today. Though, I am having a lot of issues getting screenshots instead of text from people... Really makes it fun to reference, or hope I encounter trying to reproduce. 

image

**Comment by smrhein** [comment]
Is there any help needed in getting this merged?  What needs to be investigated?

**Comment by d3v-one** [comment]
There is an error when omitting autocast("cuda") which should now be possible:

```
Traceback (most recent call last):
  File "C:\myfiles\Dev\python\diffusion\sd.py", line 352, in <module>
    image = pipe(
  File "D:\Dev\python\python-3.10\lib\site-packages\torch\autograd\grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "D:\Dev\python\python-3.10\lib\site-packages\diffusers\pipelines\stable_diffusion\pipeline_stable_diffusion.py", line 279, in __call__
    noise_pred = self.unet(latent_model_input, t, encoder_hidden_states=text_embeddings).sample
  File "D:\Dev\python\python-3.10\lib\site-packages\torch\nn\modules\module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "D:\Dev\python\python-3.10\lib\site-packages\diffusers\models\unet_2d_condition.py", line 272, in forward
    sample = self.conv_in(sample)
  File "D:\Dev\python\python-3.10\lib\site-packages\torch\nn\modules\module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "D:\Dev\python\python-3.10\lib\site-packages\torch\nn\modules\conv.py", line 457, in forward
    return self._conv_forward(input, self.weight, self.bias)
  File "D:\Dev\python\python-3.10\lib\site-packages\torch\nn\modules\conv.py", line 453, in _conv_forward
    return F.conv2d(input, weight, bias, self.stride,
RuntimeError: Input type (torch.cuda.FloatTensor) and weight type (torch.cuda.HalfTensor) should be the same


```

**Comment by shirayu** [comment]
@WASasquatch @d3v-one Is diffusers the HEAD of this Pull Request? (302a69a)
I could not reproduce them.
I think they probably have no relation to this PR.

**Comment by d3v-one** [comment]
@shirayu
The error occurs when you init the pipeline with `revision="fp16", torch_dtype=torch.float16` and omit `with autocast("cuda")`.
The `autocast` was mandatory in this case until about 3 days ago, but isn't anymore, speeding up the image generation.

**Comment by shirayu** [comment]
@d3v-one I think that is related not to THIS pull request but to #371 .
Tough that is important and should be fixed, discussed in #371 or #511.

@patrickvonplaten @patil-suraj 
Sorry to urge you, but I would prefer if you could merge this.
Thank you.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 5]
```diff
@@ -217,9 +220,25 @@ def __call__(
         do_classifier_free_guidance = guidance_scale > 1.0
         # get unconditional embeddings for classifier free guidance
         if do_classifier_free_guidance:
+            ucond_tokens: List[str]
+            if negative_prompt is None:
+                ucond_tokens = [""] * batch_size
+            elif type(prompt) is not type(negative_prompt):
+                raise TypeError("`negative_prompt` should be the same type to `prompt`.")
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:227]:
```suggestion
                raise TypeError("`negative_prompt` should be the same type to `prompt`, but got {type(negative_prompt)} != {type(prompt)}.")
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 6]
```diff
@@ -258,9 +261,25 @@ def __call__(
...
         if do_classifier_free_guidance:
+            ucond_tokens: List[str]
+            if negative_prompt is None:
+                ucond_tokens = [""] * batch_size
+            elif type(prompt) is not type(negative_prompt):
+                raise TypeError("`negative_prompt` should be the same type to `prompt`.")
+            elif isinstance(negative_prompt, str):
+                ucond_tokens = [negative_prompt] * batch_size
+            elif batch_size != len(negative_prompt):
+                raise ValueError("The length of `negative_prompt` should be equal to batch_size.")
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:272]:
```suggestion
                raise ValueError(f"`negative_prompt`: {negative_prompt} has batch size {len(negative_prompt)}, but `prompt`: {prompt} has batch size {batch_size}. Please make sure that passed `negative_prompt` matches the batch size of `prompt`.")
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py [hunk 7]
```diff
@@ -292,9 +295,25 @@ def __call__(
...
         if do_classifier_free_guidance:
+            ucond_tokens: List[str]
+            if negative_prompt is None:
+                ucond_tokens = [""] * batch_size
+            elif type(prompt) is not type(negative_prompt):
+                raise TypeError("`negative_prompt` should be the same type to `prompt`.")
+            elif isinstance(negative_prompt, str):
+                ucond_tokens = [negative_prompt] * batch_size
+            elif batch_size != len(negative_prompt):
+                raise ValueError("The length of `negative_prompt` should be equal to batch_size.")
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:306]:
```suggestion
                raise ValueError(f"`negative_prompt`: {negative_prompt} has batch size {len(negative_prompt)}, but `prompt`: {prompt} has batch size {batch_size}. Please make sure that passed `negative_prompt` matches the batch size of `prompt`.")
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_onnx.py [hunk 8]
```diff
@@ -102,9 +103,25 @@ def __call__(
...
         if do_classifier_free_guidance:
+            ucond_tokens: List[str]
+            if negative_prompt is None:
+                ucond_tokens = [""] * batch_size
+            elif type(prompt) is not type(negative_prompt):
+                raise TypeError("`negative_prompt` should be the same type to `prompt`.")
+            elif isinstance(negative_prompt, str):
+                ucond_tokens = [negative_prompt] * batch_size
+            elif batch_size != len(negative_prompt):
+                raise ValueError("The length of `negative_prompt` should be equal to batch_size.")
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_onnx.py:114]:
```suggestion
                raise ValueError(f"`negative_prompt`: {negative_prompt} has batch size {len(negative_prompt)}, but `prompt`: {prompt} has batch size {batch_size}. Please make sure that passed `negative_prompt` matches the batch size of `prompt`.")
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 9]
```diff
@@ -258,9 +261,25 @@ def __call__(
         do_classifier_free_guidance = guidance_scale > 1.0
         # get unconditional embeddings for classifier free guidance
         if do_classifier_free_guidance:
+            ucond_tokens: List[str]
+            if negative_prompt is None:
+                ucond_tokens = [""] * batch_size
+            elif type(prompt) is not type(negative_prompt):
+                raise TypeError("`negative_prompt` should be the same type to `prompt`.")
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:268]:
```suggestion
                raise TypeError("`negative_prompt` should be the same type to `prompt`, but got {type(negative_prompt)} != {type(prompt)}.")
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py [hunk 10]
```diff
@@ -292,9 +295,25 @@ def __call__(
         do_classifier_free_guidance = guidance_scale > 1.0
         # get unconditional embeddings for classifier free guidance
         if do_classifier_free_guidance:
+            ucond_tokens: List[str]
+            if negative_prompt is None:
+                ucond_tokens = [""] * batch_size
+            elif type(prompt) is not type(negative_prompt):
+                raise TypeError("`negative_prompt` should be the same type to `prompt`.")
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:302]:
```suggestion
                raise TypeError("`negative_prompt` should be the same type to `prompt`, but got {type(negative_prompt)} != {type(prompt)}.")
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_onnx.py [hunk 11]
```diff
@@ -102,9 +103,25 @@ def __call__(
         do_classifier_free_guidance = guidance_scale > 1.0
         # get unconditional embeddings for classifier free guidance
         if do_classifier_free_guidance:
+            ucond_tokens: List[str]
+            if negative_prompt is None:
+                ucond_tokens = [""] * batch_size
+            elif type(prompt) is not type(negative_prompt):
+                raise TypeError("`negative_prompt` should be the same type to `prompt`.")
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_onnx.py:110]:
```suggestion
                raise TypeError("`negative_prompt` should be the same type to `prompt`, but got {type(negative_prompt)} != {type(prompt)}.")
```

**Review by patrickvonplaten — APPROVED** [review]
Went through it again and it looks very nice! Thanks so much for working on this :-) 

Added some suggestions to improve the error messages - would be nice if you could have a look and then potentially run `make style` once if the test fails.

**Comment by patrickvonplaten** [comment]
@patil-suraj @pcuenca - I agree with @shirayu that we should merge this soon

**Comment by patrickvonplaten** [comment]
Also we would need a test here, but I'm happy to take care of this when looking into: https://github.com/huggingface/diffusers/issues/551

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 12]
```diff
@@ -217,9 +220,25 @@ def __call__(
         do_classifier_free_guidance = guidance_scale > 1.0
         # get unconditional embeddings for classifier free guidance
         if do_classifier_free_guidance:
+            ucond_tokens: List[str]
+            if negative_prompt is None:
+                ucond_tokens = [""] * batch_size
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:225]:
```suggestion
            uncond_tokens: List[str]
            if negative_prompt is None:
                uncond_tokens = [""] * batch_size
```

Minor typo.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 13]
```diff
@@ -217,9 +220,25 @@ def __call__(
...
+            elif type(prompt) is not type(negative_prompt):
+                raise TypeError("`negative_prompt` should be the same type to `prompt`.")
+            elif isinstance(negative_prompt, str):
+                ucond_tokens = [negative_prompt] * batch_size
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:232]:
```suggestion
                uncond_tokens = [negative_prompt] * batch_size
```

Same as above :)

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 14]
```diff
@@ -217,9 +220,25 @@ def __call__(
...
+            elif batch_size != len(negative_prompt):
+                raise ValueError("The length of `negative_prompt` should be equal to batch_size.")
+            else:
+                ucond_tokens = negative_prompt
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:240]:
```suggestion
                uncond_tokens = negative_prompt
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 15]
```diff
@@ -217,9 +220,25 @@ def __call__(
...
+
             max_length = text_input_ids.shape[-1]
             uncond_input = self.tokenizer(
-                [""] * batch_size, padding="max_length", max_length=max_length, return_tensors="pt"
+                ucond_tokens,
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:244]:
```suggestion
                uncond_tokens,
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 16]
```diff
@@ -258,9 +261,25 @@ def __call__(
         do_classifier_free_guidance = guidance_scale > 1.0
         # get unconditional embeddings for classifier free guidance
         if do_classifier_free_guidance:
+            ucond_tokens: List[str]
+            if negative_prompt is None:
+                ucond_tokens = [""] * batch_size
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:266]:
```suggestion
            uncond_tokens: List[str]
            if negative_prompt is None:
                uncond_tokens = [""] * batch_size
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 17]
```diff
@@ -258,9 +261,25 @@ def __call__(
...
+            elif type(prompt) is not type(negative_prompt):
+                raise TypeError("`negative_prompt` should be the same type to `prompt`.")
+            elif isinstance(negative_prompt, str):
+                ucond_tokens = [negative_prompt] * batch_size
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:273]:
```suggestion
                uncond_tokens = [negative_prompt] * batch_size
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 18]
```diff
@@ -258,9 +261,25 @@ def __call__(
...
+            elif batch_size != len(negative_prompt):
+                raise ValueError("The length of `negative_prompt` should be equal to batch_size.")
+            else:
+                ucond_tokens = negative_prompt
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:277]:
```suggestion
                uncond_tokens = negative_prompt
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 19]
```diff
@@ -258,9 +261,25 @@ def __call__(
...
+
             max_length = text_input_ids.shape[-1]
             uncond_input = self.tokenizer(
-                [""] * batch_size, padding="max_length", max_length=max_length, return_tensors="pt"
+                ucond_tokens,
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:281]:
```suggestion
                uncond_tokens,
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py [hunk 20]
```diff
@@ -292,9 +295,25 @@ def __call__(
         do_classifier_free_guidance = guidance_scale > 1.0
         # get unconditional embeddings for classifier free guidance
         if do_classifier_free_guidance:
+            ucond_tokens: List[str]
+            if negative_prompt is None:
+                ucond_tokens = [""] * batch_size
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:300]:
```suggestion
            uncond_tokens: List[str]
            if negative_prompt is None:
                uncond_tokens = [""] * batch_size
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py [hunk 21]
```diff
@@ -292,9 +295,25 @@ def __call__(
...
+            elif type(prompt) is not type(negative_prompt):
+                raise TypeError("`negative_prompt` should be the same type to `prompt`.")
+            elif isinstance(negative_prompt, str):
+                ucond_tokens = [negative_prompt] * batch_size
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:307]:
```suggestion
                uncond_tokens = [negative_prompt] * batch_size
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py [hunk 22]
```diff
@@ -292,9 +295,25 @@ def __call__(
...
+            elif batch_size != len(negative_prompt):
+                raise ValueError("The length of `negative_prompt` should be equal to batch_size.")
+            else:
+                ucond_tokens = negative_prompt
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:315]:
```suggestion
                uncond_tokens = negative_prompt
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py [hunk 23]
```diff
@@ -292,9 +295,25 @@ def __call__(
...
+
             max_length = text_input_ids.shape[-1]
             uncond_input = self.tokenizer(
-                [""] * batch_size, padding="max_length", max_length=max_length, return_tensors="pt"
+                ucond_tokens,
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:319]:
```suggestion
                uncond_tokens,
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_onnx.py [hunk 24]
```diff
@@ -102,9 +103,25 @@ def __call__(
         do_classifier_free_guidance = guidance_scale > 1.0
         # get unconditional embeddings for classifier free guidance
         if do_classifier_free_guidance:
+            ucond_tokens: List[str]
+            if negative_prompt is None:
+                ucond_tokens = [""] * batch_size
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_onnx.py:108]:
```suggestion
            uncond_tokens: List[str]
            if negative_prompt is None:
                uncond_tokens = [""] * batch_size
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_onnx.py [hunk 25]
```diff
@@ -102,9 +103,25 @@ def __call__(
...
+            elif type(prompt) is not type(negative_prompt):
+                raise TypeError("`negative_prompt` should be the same type to `prompt`.")
+            elif isinstance(negative_prompt, str):
+                ucond_tokens = [negative_prompt] * batch_size
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_onnx.py:115]:
```suggestion
                uncond_tokens = [negative_prompt] * batch_size
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_onnx.py [hunk 26]
```diff
@@ -102,9 +103,25 @@ def __call__(
...
+            elif batch_size != len(negative_prompt):
+                raise ValueError("The length of `negative_prompt` should be equal to batch_size.")
+            else:
+                ucond_tokens = negative_prompt
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_onnx.py:123]:
```suggestion
                uncond_tokens = negative_prompt
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_onnx.py [hunk 27]
```diff
@@ -102,9 +103,25 @@ def __call__(
...
+
             max_length = text_input_ids.shape[-1]
             uncond_input = self.tokenizer(
-                [""] * batch_size, padding="max_length", max_length=max_length, return_tensors="np"
+                ucond_tokens,
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_onnx.py:127]:
```suggestion
                uncond_tokens,
```

**Comment by shirayu** [comment]
@patrickvonplaten Thank you for your nice updates!
I merged them and fix styles.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 28]
```diff
@@ -144,6 +145,8 @@ def __call__(
                 Paper](https://arxiv.org/pdf/2205.11487.pdf). Guidance scale is enabled by setting `guidance_scale >
                 1`. Higher guidance scale encourages to generate images that are closely linked to the text `prompt`,
                 usually at the expense of lower image quality.
+            negative_prompt (`str` or `List[str]`, *optional*):
+                The prompt or prompts not to guide the image generation.
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:149]:
```suggestion
                The prompt or prompts not to guide the image generation. Ignored when not using guidance (i.e., ignored if `guidance_scale` is not greater than `1`).
```

Same for the other pipelines, if you agree :)

**Review by pcuenca — COMMENTED** [review]
The code looks good, I just suggested minor nits. I haven't been able to test it yet, I'll do it later.

I have a question about how it works, though. Instead of using an empty prompt for the unconditioned generation we use the negative prompt, is that correct? Wouldn't this make those prompts appear for low values of `guidance_scale`?

**Comment by shirayu** [comment]
@pcuenca Thank you for your suggestions. I update this PR.

As far as I know, in almost all cases `guidance_scale` greater than 1 like 7.5.
FYI: [This article](https://www.reddit.com/r/StableDiffusion/comments/x41n87/how_to_get_images_that_dont_suck_a/) is a note about the CFG value I found before.
```
CFG 2 - 6: Let the AI take the wheel.
CFG 7 - 11: Let's collaborate, AI!
CFG 12 - 15: No, seriously, this is a good prompt. Just do what I say, AI.
CFG 16 - 20: DO WHAT I SAY OR ELSE, AI.
```

Replacing blanks in unconditional generation is done in some applications like `AUTOMATIC1111//stable-diffusion-webui`, though I can't figure out a paper to shows its theoretical validity.

https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/3ff0de2c594b786ef948a89efb1814c59bb42117/modules/processing.py#L354

**Comment by shirayu** [comment]
> Shouldn't there be checks (printing a warning) for the maximum number of tokens in the negative prompt as well?

Good suggestion!
However, it becomes more complex PR and the code will be repeat of previous operations.
So, I think it should be resolved by another Pull Request after #551 resolved.

(Related to #472)

**Comment by allo-** [comment]
I guess that's the tradeoff between separate pipelines which can easily be copied to create a own one and an architecture that can do much more but maybe harder to understand at first. In #699 and the Twitter poll it didn't look like there will be a unified pipeline soon.

But maybe the tokenizing (with the warning when there are too many tokens) could be moved to a function (maybe even into one of the utility files?) so it can be reused for both prompts and in all the Pipelines.

**Review by patil-suraj — APPROVED** [review]
Looks good to me, thanks a lot for working on this!

**Pull request merged**